crates/ruff_python_ast/src/source_code/line_index.rs

```diff
@@ -236,49 +236,71 @@ impl IndexKind {
 /// Type-safe wrapper for a value whose logical range starts at `1`, for
 /// instance the line or column numbers in a file
 ///
-/// Internally this is represented as a [`NonZeroUsize`], this enables some
+/// Internally this is represented as a [`NonZeroU32`], this enables some
 /// memory optimizations
+///
+/// Conversion between `usize` and `OneIndexed` always specify
+/// zero-indexed or one-indexed. The internal representation is always
+/// one-indexed.
 #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
-pub struct OneIndexed(NonZeroUsize);
+pub struct OneIndexed(NonZeroU32);
+
+#[allow(clippy::cast_possible_truncation)] // manually checked
+const fn index_to_u32(value: usize) -> u32 {
+    assert!(
+        value <= u32::MAX as usize,
+        "Value is too large to be represented as a OneIndexed"
+    );
+    value as u32
+}
 
 impl OneIndexed {
     // SAFETY: These constants are being initialized with non-zero values
     /// The smallest value that can be represented by this integer type.
     pub const MIN: Self = unwrap(Self::new(1));
     /// The largest value that can be represented by this integer type
-    pub const MAX: Self = unwrap(Self::new(usize::MAX));
+    pub const MAX: Self = unwrap(Self::new(u32::MAX));
 
-    pub const ONE: NonZeroUsize = unwrap(NonZeroUsize::new(1));
+    pub const ONE: NonZeroU32 = unwrap(NonZeroU32::new(1));
 
     /// Creates a non-zero if the given value is not zero.
-    pub const fn new(value: usize) -> Option<Self> {
-        match NonZeroUsize::new(value) {
+    pub const fn new(value: u32) -> Option<Self> {
+        match NonZeroU32::new(value) {
             Some(value) => Some(Self(value)),
             None => None,
         }
     }
 
     /// Construct a new [`OneIndexed`] from a zero-indexed value
+    ///
+    /// # Panics
+    /// Panics if the given value doesn't fit in u32
     pub const fn from_zero_indexed(value: usize) -> Self {
```

Comment:
I agree making this function to take u32 is better.

About `to_one_indexed`, I will follow your decision, but I'd like to add one more comment.

Unlike `NonZero` type, which means zero is out of range, `OneIndexed` offers conversion functions to zero indexed value. I also can say `OneIndexed` is another representation of zero indexed value.
When user convert this value to usize, sometimes it is conversion to underlying value (zero-indexed) and something it is conversion to representation (one-indexed).
By providing two symmetric interfaces, users can always select their intentions with equal weight by looking at the function names. It helps to reduce mistakes.
By providing `From<OneIndexed> for usize`, users tends to regards one way is default behavior and the other way is optional function.  It may be the truth because this type is designed to hold one-indexed values. But optional function is easy to be forgotten.
Practically it might be confusing. It will not be confused in one way - using to_zero_indexed is very explicit, but the other way - writing `value.into()` for `usize` conversion - is very easy to happen inconsiously.

Ruff codebase is using either conversion in similar ratio in small numbers - to-zero 8 and to-one 7. Implementing `From<OneIndexed> for usize` doesn't give much value and also hard to say which conversion is dominant. I personally will not provide that `From` implementation while it supports both one and zero indexed conversions.

Reply:
> About to_one_indexed, I will follow your decision, but I'd like to add one more comment.

That's an excellent point. I guess what I find confusing is that both `get` and `to_one_indexed` return the same value and only differ by the return type. What do you think of `get` and `get_usize()` (or `to_one_indexed()` and `to_one_indexed_usize()`?

Reply:
Otherwise, just full matrix also makes sense to me.

`to_one_indexed() -> u32`
`to_zero_indexed() -> u32`
`to_one_indexed_usize() -> usize`
`to_zero_indexed_usize() -> usize`

Reply:
Otherwise, just full matrix also makes sense to me.

`to_one_indexed() -> u32`
`to_zero_indexed() -> u32`
`to_one_indexed_usize() -> usize`
`to_zero_indexed_usize() -> usize`